File: python/tradingflow/predictor/mean/sample.py
import numpy as np

from ._base import MeanPredictor


def fit(x: np.ndarray, y: np.ndarray) -> np.ndarray:
    """Per-stock sample mean of the target over the window, ignoring features."""
    return np.nanmean(y, axis=0)


def predict(features: np.ndarray, params: np.ndarray) -> np.ndarray:
    return params


def build(*, retain_features: bool = False, **kwargs) -> MeanPredictor:
    """Constructs a sample-mean predictor — the do-nothing baseline every other
    mean predictor has to beat.

    It fits from the target alone, so it retains no feature history and does
    not withhold a stock whose features are unusable.
    """
    return MeanPredictor(
        fit=fit,
        predict=predict,
        retain_features=retain_features,
        **kwargs,
    )

File: crates/tradingflow/src/operators/predictor/mean/sample.rs
use super::MeanPredictor;
use crate::operators::predictor::Config;
use crate::python::py_segment_module;

/// Predicts each stock's return as its own sample mean over the window,
/// ignoring features entirely.
///
/// The baseline every feature-driven predictor has to beat: it captures
/// whatever persistent cross-sectional drift exists and nothing else, so a
/// model that cannot outperform it is not extracting signal from its features.
///
/// See [module-level docs](super::super) for inputs and outputs.
pub fn sample(config: Config) -> impl MeanPredictor {
    py_segment_module(
        "tradingflow.predictor.mean.sample",
        config.params(|_| Ok(())),
    )
}
